File: examples/hpc/omp/adder.f90
program hello_world
use omp_lib
implicit none
integer :: i, n
real(8) :: s, s_tmp, t
t = omp_get_wtime()
!$omp parallel private(i, s_tmp) &
!$omp   shared(s)
s = 0.0_8
s_tmp = 0.0_8
n = 2e7
!$omp do
do i = 1, n
  s_tmp = s_tmp + dble(i)
end do
!$omp end do
!$omp atomic
s = s + s_tmp
!$omp end parallel
t = omp_get_wtime() - t
write (*,"(f16.0, es10.2)") s, t
end program hello_world

File: examples/hpc/omp/adder.cc
#include <omp.h>
#include <cstdio>
using std::printf;
int main()
{
  int n = 2e7;
  double s = 0.0, s_tmp;
  double t0 = omp_get_wtime(), te;
#pragma omp parallel \
   private(s_tmp) shared(n, s)
{
  s_tmp = 0.0;
#pragma omp for
  for (int i = 0; i < n; ++i)
  {
    s_tmp = s_tmp + i;
  }
#pragma omp atomic
  s = s + s_tmp;
} // end parallel
te = omp_get_wtime() - t0;
printf("%16.0f, et=%10.2e\n", s, te);
return 0;
}
